Add system_group option

takumin opened the pull request:
## Summary

Add a `system_group` attribute to the `group` resource.
It is forwarded to specinfra's `add_group`, which turns it into `-r`/`--system` of `groupadd(8)`.

```ruby
group "create itamae_system group" do
  groupname "itamae_system"
  system_group true
end
```

## Motivation

The `user` resource accepts `system_user`.
The `group` resource has no counterpart, so a recipe cannot create a system group.

## Changes

- `group` resource: accept `system_group` and pass it to `add_group`.
- Bump `specinfra` from `>= 2.64.0` to `>= 2.81.0`.
  - v2.80.0 added `system_group` to `add_group` (mizzy/specinfra@216f731).
  - Older versions silently ignore it and create a regular group.
  - v2.81.0 added `check_is_system_group` / `check_is_system_user`
    (mizzy/specinfra@990dbc9), used by the integration specs.
- Bump `serverspec` from `~> 2.1` to `~> 2.43`.
  - v2.43.0 added the `is_system_group?` / `is_system_user?` matchers
    (mizzy/serverspec@0cd1cb9). They do not exist in 2.1.
- Integration tests: create an `itamae_system` group and user, with `system_group true` and `system_user true` respectively.
  - The user's primary group is the `itamae_system` group.
  - Verified with `be_is_system_group` / `be_is_system_user`.
  - `system_user` had no integration coverage before either.

takumin committed:
Add system_group option

- `-r`, `--system` of `groupadd(8)` creates a system group
  - Passed to specinfra `add_group` as `system_group`
  - Requires specinfra >= 2.81.0
- The `user` resource already accepts `system_user`
  - The `group` resource had no counterpart

takumin commented:
Merged Specinfra: https://github.com/mizzy/specinfra/commit/30e1c2177f7b8a2b8e2a8302a8492373480b3641

unasuke reviewed (approved):
Seems good! Could you add any test case if possible?

takumin commented:
thank you for your comment.
I'm sorry I forgot to add a test case.

I think Serverspec needs a new resource type.
Add a test case when a new resource type is added.

takumin commented:
I created the following pull request:

https://github.com/mizzy/specinfra/pull/694

Besides this, I would like to add test cases after the necessary corrections are merged.

It takes a long time, I am sorry.

unasuke commented:
Great!!

unasuke commented:
Sorry, I forgot about it.

takumin commented:
I'm so so so... so sorry for making you wait so long.
I just submitted a PR to serverspec, so I'll add some test code once the above is merged.

takumin committed:
Add integration tests for system_user and system_group

- `itamae_system` group is created with `system_group true`
- `itamae_system` user is created with `system_user true`
  - Its primary group is the `itamae_system` group
- Verified with `be_is_system_group` / `be_is_system_user`
- Neither option had any integration coverage

takumin commented:
@unasuke 
This PR passed the test.

```
User "itamae"
  is expected to exist
  is expected to have uid 1234
  is expected to have home directory "/home/itamae"
  is expected to have login shell "/bin/dash"
  is expected not to be is system user

User "itamae_system"
  is expected to exist
  is expected to be is system user

Group "itamae_system"
  is expected to exist
  is expected to be is system group
```

However, an error has occurred that is unrelated to this PR.

```
Pending: (Failures listed here are expected and do not affect your suite's status)

  1) File "/tmp/http_request_redirect.html"
     # Temporarily skipped with xdescribe
     # ./spec/integration/default_spec.rb:119

  2) File "/tmp/http_request_redirect.html" content
     # Temporarily skipped with xdescribe
     # ./spec/integration/default_spec.rb:120


Failures:

  1) File "/tmp/file_edit_with_content_change_updates_timestamp" mtime is expected to be > 2016-05-02T01:23:45+00:00
     Failure/Error: its(:mtime) { should be > DateTime.iso8601("2016-05-02T01:23:45Z") }
       expected: > #<DateTime: 2016-05-02T01:23:45+00:00 ((2457511j,5025s,0n),+0s,2299161j)>
            got:   #<DateTime: 1980-01-02T09:00:00+09:00 ((2444241j,0s,0n),+32400s,2299161j)>

     # ./spec/integration/default_spec.rb:287:in 'block (2 levels) in <top (required)>'

  2) File "/tmp/file_with_content_change_updates_timestamp" mtime is expected to be > 2016-05-01T01:23:45+00:00
     Failure/Error: its(:mtime) { should be > DateTime.iso8601("2016-05-01T01:23:45Z") }
       expected: > #<DateTime: 2016-05-01T01:23:45+00:00 ((2457510j,5025s,0n),+0s,2299161j)>
            got:   #<DateTime: 1980-01-02T09:00:00+09:00 ((2444241j,0s,0n),+32400s,2299161j)>

     # ./spec/integration/default_spec.rb:319:in 'block (2 levels) in <top (required)>'

Finished in 3.06 seconds (files took 0.23664 seconds to load)
153 examples, 2 failures, 2 pending

Failed examples:

rspec ./spec/integration/default_spec.rb:287 # File "/tmp/file_edit_with_content_change_updates_timestamp" mtime is expected to be > 2016-05-02T01:23:45+00:00
rspec ./spec/integration/default_spec.rb:319 # File "/tmp/file_with_content_change_updates_timestamp" mtime is expected to be > 2016-05-01T01:23:45+00:00
```

I don't know the cause, could you please help me... sorry...

takumin commented:
I ran it three times in my local environment with version 3.4 and confirmed that the same error occurred every time.

takumin commented:
Let's take a look inside the container.

```console
$ docker exec itamae sh -c 'ls -la /tmp/file_edit_with_content_change_updates_timestamp'
-rw-r--r-- 1 root root 14 Jan  2  1980 /tmp/file_edit_with_content_change_updates_timestamp
$ docker exec itamae sh -c 'ls -la /tmp/file_with_content_change_updates_timestamp'
-rw------- 1 root root 12 Jan  2  1980 /tmp/file_with_content_change_updates_timestamp
```

I'll try updating the date.

```console
$ docker exec itamae sh -c 'touch -d 2016-05-02T12:34:56Z /tmp/file_edit_with_content_change_updates_timestamp'
$ docker exec itamae sh -c 'touch -d 2016-05-02T12:34:56Z /tmp/file_with_content_change_updates_timestamp'
```

Let's check inside the container again.

```console
$ docker exec itamae sh -c 'ls -la /tmp/file_edit_with_content_change_updates_timestamp'
-rw-r--r-- 1 root root 14 May  2  2016 /tmp/file_edit_with_content_change_updates_timestamp
$ docker exec itamae sh -c 'ls -la /tmp/file_with_content_change_updates_timestamp'
-rw------- 1 root root 12 May  2  2016 /tmp/file_with_content_change_updates_timestamp
```

Updating the date using the `touch` command seems to be no problem.

takumin commented:
This is the code where I think the problem is occurring.

https://github.com/itamae-kitchen/itamae/blob/57dfb50d93c836e77911bf3a592be8bd56ec21ba/spec/integration/recipes/default.rb#L458-L465

https://github.com/itamae-kitchen/itamae/blob/57dfb50d93c836e77911bf3a592be8bd56ec21ba/spec/integration/recipes/default.rb#L498-L502

https://github.com/itamae-kitchen/itamae/blob/57dfb50d93c836e77911bf3a592be8bd56ec21ba/spec/integration/default_spec.rb#L275-L277

https://github.com/itamae-kitchen/itamae/blob/57dfb50d93c836e77911bf3a592be8bd56ec21ba/spec/integration/default_spec.rb#L307-L309

takumin commented:
@unasuke @sue445 
I was able to confirm that this bug is unrelated to this PR, so I have registered an issue and prepared a repository for reproduction.

I would like to use this PR for work, so I would like it to be merged. Is that possible?

takumin commented:
@sue445 @unasuke

I have rebased this onto the latest master and added the integration test cases that were requested:

- `group "itamae_system"` with `system_group true` and `user "itamae_system"` with `system_user true` in `spec/integration/recipes/default.rb`
- `be_is_system_group` / `be_is_system_user` expectations in `spec/integration/default_spec.rb`

The unrelated mtime failure I reported earlier (#377) has since been fixed, and CI is fully green on this branch now.

One thing I would like your opinion on before merging: the specinfra side of `system_group` landed in specinfra v2.80.0 (mizzy/specinfra@30e1c2177f7b8a2b8e2a8302a8492373480b3641), but `itamae.gemspec` currently allows `specinfra >= 2.64.0`. With specinfra 2.64-2.79 the `system_group` attribute would be silently ignored rather than raising an error. Would you like me to bump the minimum to `>= 2.80.0` in this PR, or leave the constraint as is and handle it separately?

Could you take another look when you have time? Thanks!

takumin committed:
Require specinfra >= 2.81.0

- `groupadd -r` creates a system group
  - Used by the `group` resource, via `add_group`
  - Landed in specinfra v2.80.0
  - mizzy/specinfra@216f731
- `check_is_system_group` / `check_is_system_user` verify one
  - Used by the integration specs
  - Landed in specinfra v2.81.0
  - mizzy/specinfra@990dbc9
- The gemspec allowed `>= 2.64.0`
  - system_group is accepted but silently ignored there

takumin committed:
Require serverspec >= 2.43.0

- `is_system_group?` / `is_system_user?` verify a system account
  - Used by the integration specs
  - Landed in serverspec v2.43.0
  - mizzy/serverspec@0cd1cb9
- The gemspec allowed `~> 2.1`
  - The matchers do not exist there

unasuke commented:
@takumin Sorry, late. This is a hard question. My concern is whether raising the required version of specinfra will also necessitate changing the minimum required Ruby version that itamae demands.

* https://rubygems.org/gems/specinfra/versions/2.64.0
* https://rubygems.org/gems/specinfra/versions/2.80.0

Both the current minimum specinfra gem required by itamae and the version that supports `system_group` support have no restrictions on Ruby versions. Additionally, we've confirmed that the latest release, specinfra v2.95.1, runs its CI on sufficiently old Ruby 2.3.

https://github.com/mizzy/specinfra/actions/runs/28725810470

Therefore, I believe raising the minimum version requirement to `>= 2.80.0` is appropriate, and in some scenarios, even removing the version constraint altogether might be preferable.

However, allowing the system to silently ignore the requirement could lead to user confusion, so setting the specinfra requirement to at least version 2.80.0 would be the more prudent choice.

I hope @sue445 shares the same idea - what do you think?

sue445 commented:
@unasuke 

> Therefore, I believe raising the minimum version requirement to `>= 2.80.0` is appropriat

The patch currently specifies `>= 2.81.0`. Is `>= 2.80.0` a typo?

In any case, since version `2.81.0` was released back in 2019 [^1], I believe raising the minimum version to `2.81.0` won't cause any major issues.

[^1]: https://rubygems.org/gems/specinfra/versions/2.81.0

takumin commented:
@unasuke @sue445 Thank you both!

The `2.80.0` vs `2.81.0` discrepancy is my fault, not a typo by @unasuke — sorry for the confusion. When I asked the question above, the branch still had `>= 2.64.0` and I proposed `>= 2.80.0`, where the creation side of the feature landed (mizzy/specinfra@30e1c21). Afterwards, while reorganizing the commits, I updated the gemspec to `>= 2.81.0`, since that release ships the `check_is_system_group` / `check_is_system_user` commands backing the `be_is_system_group` / `be_is_system_user` matchers used by the integration tests in this PR (mizzy/specinfra#694). Apart from the version bump, the diff between the two releases is exactly that change: https://github.com/mizzy/specinfra/compare/v2.80.0...v2.81.0

@unasuke Thank you for checking the Ruby version implications — good to know that neither the 2.64.0 nor the 2.80.0 gemspec constrains the Ruby version, and that the latest specinfra still runs its CI on Ruby 2.3. I also agree that silently ignoring `system_group` on older specinfra would be more confusing for users than a stricter constraint.

Since @sue445 confirmed `>= 2.81.0` won't cause any issues, I'll keep the constraint as is. Thanks!

takumin commented:
Thank you so much for merging this!

It has been just over seven years since I opened this PR, so I’m genuinely happy to see it finally land.

Thank you, @unasuke and @sue445, for revisiting it, reviewing the changes, and helping resolve the dependency version question. I really appreciate your continued work on Itamae!
